examples/step-20/step-20.cc

```diff
@@ -559,206 +566,86 @@ namespace Step20
+  // more, but rather only comment on some remaining implementational
+  // aspects.
 
+  // @sect4{MixedLaplace::solve}
 
-  void SchurComplement::vmult(Vector<double> &      dst,
-                              const Vector<double> &src) const
+  // As already outlined in the introduction, the solve function consists
+  // essentially of two steps. First, we have to form the first equation
+  // involving the Schur complement and solve for the pressure (component 1
+  // of the solution). Then, we can reconstruct the velocities from the
+  // second equation (component 0 of the solution).
+  template <int dim>
+  void MixedLaplaceProblem<dim>::solve()
   {
-    system_matrix->block(0, 1).vmult(tmp1, src);
-    m_inverse->vmult(tmp2, tmp1);
-    system_matrix->block(1, 0).vmult(dst, tmp2);
-  }
+    // As a first step we declare references to all block components of the
+    // matrix, the right hand side and the solution vector that we will
+    // need.
+    const auto &M = system_matrix.block(0, 0);
+    const auto &B = system_matrix.block(0, 1);
 
+    const auto &F = system_rhs.block(0);
+    const auto &G = system_rhs.block(1);
 
-  // @sect4{The <code>ApproximateSchurComplement</code> class}
+    auto &U = solution.block(0);
+    auto &P = solution.block(1);
 
-  // The third component of our solver and preconditioner system is the class
-  // that approximates the Schur complement with the method described in the
-  // introduction. We will use this class to build a preconditioner for our
-  // system matrix.
-  class ApproximateSchurComplement : public Subscriptor
-  {
-  public:
-    ApproximateSchurComplement(const BlockSparseMatrix<double> &A);
+    // Then, we will create corresponding LinearOperator objects and create
+    // the <code>op_M_inv</code> operator:
+    const auto op_M = linear_operator(M);
+    const auto op_B = linear_operator(B);
 
-    void vmult(Vector<double> &dst, const Vector<double> &src) const;
+    ReductionControl     reduction_control_M(2000, 1.0e-18, 1.0e-10);
+    SolverCG<>           solver_M(reduction_control_M);
+    PreconditionJacobi<> preconditioner_M;
 
-  private:
-    const SmartPointer<const BlockSparseMatrix<double>> system_matrix;
+    preconditioner_M.initialize(M);
 
-    mutable Vector<double> tmp1, tmp2;
-  };
+    const auto op_M_inv = inverse_operator(op_M, solver_M, preconditioner_M);
 
+    // This puts us in the position to be able to declare the Schur
+    // complement <code>op_S</code> and the approximate Schur complement
+    // <code>op_aS</code>:
+    const auto op_S = transpose_operator(op_B) * op_M_inv * op_B;
+    const auto op_aS =
+      transpose_operator(op_B) * linear_operator(preconditioner_M) * op_B;
 
+    // We now create a preconditioner out of <code>op_aS</code> that
+    // applies a few number of CG iterations (until a very modest relative
+    // reduction of $10^{-16}$ is reached):
+    ReductionControl     reduction_control_aS(2000, 1.e-18, 1.0e-6);
```

Comment:
Even though we probably required a reduction rate like `1e-6` before, I would feel that `1e-3` should be a modest tolerance at this point since we're talking about the approximation of the mass matrix by a diagonal matrix. The question is then of course whether the `15 iterations` we report in the results file is still adequate. (If not, then `1e-3` is too coarse after all.)

Reply:
It's not the approximation of the mass matrix, but the approximation of the Schur complement using a diagonal mass matrix. 

But this doesn't take anything away from your comment about trying 1e-3!

Reply:
For 5 global refinement steps a tolerance of 1e-3 has a bit better overall runtime performance than 1e-6. I will change the step to 1e-3 and adjust the results section. (After all 1e-6 was eyeballing it anyway. In order to achieve 15 iterations we would have to choose 1e-10...)

Reply:
An issue you may want to consider is that CG requires the preconditioner to be a linear operator. But `inverse_operator(M, prec, tol)` is really only a linear operator if `tol` is sufficiently smooth -- otherwise it is nonlinear and may destroy the CG convergence properties for the outer solver.

I'm not sure what tolerance is necessary, but it's worth keeping in mind if you run into errors from inside CG about failure of convergence.

Reply:
@bangerth Actually, what about I change the preconditioner to be controlled by `IterationNumberControl` and to simply fix it to say 5 CG iteratons.
That way the operator remains the same linear function.

Reply:
@bangerth Updated. Have a look at the last commit. This increased the overall runtime by a factor of 2 (for the large testcase).

Reply:
This looks great from my point of view, thanks!

Reply:
As a side note, we might at some point think if we should implement a variant of the CG method which is more robust for somewhat non-linear preconditioners, by implementing the so-called [Polak-Ribiere update formula](https://en.wikipedia.org/wiki/Conjugate_gradient_method#The_flexible_preconditioned_conjugate_gradient_method). If anyone else also thinks this would be nice to have (at least for a try, I have no idea how it actually compares), I will open an issue.

Reply:
@tamiko: No, fixed number of iterations still doesn't make it a linear operator. (It becomes a linear operator if you had something like a fixed point iteration, think Richardson or SSOR, but it's not a linear operation for CG. You can see that if you just write out the operations done for one or two iterations.)

If this increases the run-time by a factor of two, I'd be in favor of going back to the previous state.

@kronbichler: Yes, I think that would be nice to have! It's the equivalent of F-GMRES.

Reply:
@kronbichler I also think that would be a nice feature to have :-)